c2_utils/src/mfx_defs.cpp

```diff
@@ -42,57 +42,181 @@ static void InitMfxFrameHeader(
+
+        // if input surface width or height is not 16bit aligned, do copy here
+        for (i = 0; i < nCropH/2; ++i)
+        {
+            // copying Y
+            uint8_t *src = Y + nCropX + (nCropY + i)*nOPitch;
+            uint8_t *dst = srf->Data.Y + nCropX + (nCropY + i)*nPitch;
+            std::copy(src, src + nCropW, dst);
+
+            // copying UV
+            src = UV + nCropX + (nCropY/2 + i)*nOPitch;
+            dst = srf->Data.UV + nCropX + (nCropY/2 + i)*nPitch;
+            std::copy(src, src + nCropW, dst);
+        }
+        for (i = nCropH/2; i < nCropH; ++i)
+        {
+            // copying Y (remained data)
+            uint8_t *src = Y + nCropX + (nCropY + i)*nOPitch;
+            uint8_t *dst = srf->Data.Y + nCropX + (nCropY + i)*nPitch;
+            std::copy(src, src + nCropW, dst);
+        }
+    }
+
+    return res;
 }
 
-void InitMfxFrameHW(
+mfxStatus InitMfxFrameHW(
     uint64_t timestamp, uint64_t frame_index,
     mfxMemId mem_id,
     uint32_t width, uint32_t height, uint32_t fourcc, const mfxFrameInfo& info, mfxFrameSurface1* mfx_frame)
 {
     MFX_DEBUG_TRACE_FUNC;
+    mfxStatus res = MFX_ERR_NONE;
 
     InitMfxFrameHeader(timestamp, frame_index, width, height, fourcc, info, mfx_frame);
 
     MFX_ZERO_MEMORY(mfx_frame->Data);
     mfx_frame->Data.MemType = MFX_MEMTYPE_EXTERNAL_FRAME;
     mfx_frame->Data.MemId = mem_id;
+
+    return res;
+}
+
+uint32_t MFXGetSurfaceSize(uint32_t FourCC, uint32_t width, uint32_t height)
+{
+    MFX_DEBUG_TRACE_FUNC;
+
+    mfxU32 nbytes = 0;
+
+    switch (FourCC) {
+        case MFX_FOURCC_I420:
+        case MFX_FOURCC_NV12:
+            nbytes = width * height + (width >> 1) * (height >> 1) + (width >> 1) * (height >> 1);
+            break;
+        case MFX_FOURCC_I010:
+        case MFX_FOURCC_P010:
+            nbytes = width * height + (width >> 1) * (height >> 1) + (width >> 1) * (height >> 1);
+            nbytes *= 2;
+            break;
+        case MFX_FOURCC_RGB4:
+            nbytes = width * height * 4;
+            break;
+        default:
+            break;
+    }
+
+    return nbytes;
+}
+
+uint32_t MFXGetFreeSurfaceIdx(mfxFrameSurface1 *SurfacesPool, uint32_t nPoolSize)
+{
+    MFX_DEBUG_TRACE_FUNC;
+
+    for (uint32_t i = 0; i < nPoolSize; i++) {
+        if (0 == SurfacesPool[i].Data.Locked)
+            return i;
+    }
+    return MFX_ERR_NOT_FOUND;
+}
+
+mfxStatus MFXAllocSystemMemorySurfacePool(uint8_t **buf, mfxFrameSurface1 *surfpool, mfxFrameInfo frame_info, uint32_t surfnum)
+{
+    MFX_DEBUG_TRACE_FUNC;
+
+    // initialize surface pool (I420, RGB4 format)
+    if (!surfpool)
+        return MFX_ERR_NULL_PTR;
+
+    uint32_t surfaceSize = MFXGetSurfaceSize(frame_info.FourCC, frame_info.Width, frame_info.Height);
+    if (!surfaceSize)
+        return MFX_ERR_MEMORY_ALLOC;
+
+    size_t framePoolBufSize = static_cast<size_t>(surfaceSize) * surfnum;
+    *buf = reinterpret_cast<uint8_t *>(calloc(framePoolBufSize, 1));
```

Comment:
I don't think it's good idea to alloc so big buffer which cantain all surface, why don't alloc a single buffer per surface?

Reply:
It is just for input frames which are not 16byte aligned. If width and height are 16byte aligned, there is no need to allocate this buffer.  I have below statement to control this buffer allocation.

// 16-byte align requirement
#define MFX_C2_IS_COPY_NEEDED(_mem_type, _input_info, _mfx_info)\
    ((MFX_MEMTYPE_SYSTEM_MEMORY == _mem_type) && \
    (((_input_info).Width < (_mfx_info.Width)) || \
    ((_input_info).Height < (_mfx_info.Height))))